gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java

```diff
@@ -0,0 +1,202 @@
+    this.callable = Preconditions.checkNotNull(builder.unaryCallable);
+    this.batchingDescriptor = Preconditions.checkNotNull(builder.batchingDescriptor);
+  }
+
+  public static class Builder<ElementT, ElementResultT, RequestT, ResponseT> {
+    private BatchingDescriptor<ElementT, ElementResultT, RequestT, ResponseT> batchingDescriptor;
+    private UnaryCallable<RequestT, ResponseT> unaryCallable;
+    private RequestT prototype;
+
+    private Builder() {}
+
+    public Builder<ElementT, ElementResultT, RequestT, ResponseT> setBatchingDescriptor(
+        BatchingDescriptor<ElementT, ElementResultT, RequestT, ResponseT> batchingDescriptor) {
+      this.batchingDescriptor = batchingDescriptor;
+      return this;
+    }
+
+    public Builder<ElementT, ElementResultT, RequestT, ResponseT> setUnaryCallable(
+        UnaryCallable<RequestT, ResponseT> unaryCallable) {
+      this.unaryCallable = unaryCallable;
+      return this;
+    }
+
+    public Builder<ElementT, ElementResultT, RequestT, ResponseT> setPrototype(RequestT prototype) {
+      this.prototype = prototype;
+      return this;
+    }
+
+    /** Build the BatcherImpl */
+    public BatcherImpl<ElementT, ElementResultT, RequestT, ResponseT> build() {
+      return new BatcherImpl<>(this);
+    }
+  }
+
+  public static <EntryT, EntryResultT, RequestT, ResponseT>
+      Builder<EntryT, EntryResultT, RequestT, ResponseT> newBuilder() {
+    return new Builder<>();
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public ApiFuture<ElementResultT> add(final ElementT element) {
+    Preconditions.checkState(!isClosed, "Cannot perform batching on a closed connection");
+    lock.lock();
+    try {
+      if (batch == null) {
+        batch = new Batch(batchingDescriptor.newRequestBuilder(prototype));
+      }
+      SettableApiFuture<ElementResultT> result = SettableApiFuture.create();
+      batch.add(element, result);
+      ApiFutures.addCallback(
+          result,
+          new ApiFutureCallback<ElementResultT>() {
+            @Override
+            public void onFailure(Throwable throwable) {
+              onResultResolve();
+            }
+
+            @Override
+            public void onSuccess(ElementResultT elementResultT) {
+              onResultResolve();
+            }
+          },
+          directExecutor());
+      return result;
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private void onResultResolve() {
+    latch.countDown();
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void flush() throws InterruptedException {
+    lock.lock();
+    try {
+      sendBatch();
+      latch.await();
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /** sends accumulated elements asynchronously for batching. */
+  private void sendBatch() {
+    if (batch == null) {
+      return;
+    }
+    final Batch accumulatedBatch = batch;
+    latch = new CountDownLatch(accumulatedBatch.results.size());
```

Comment:
Reinitializing `latch` with the number of elements in the batch and waits till the end of batch via latch.await(). 

Does this logic seems ok.. or is there any better way to enforce synchronous `sendBatch`.